Overhaul Spirit Temple and Desert Colossus logic

Hexocyte opened the pull request:
This hardens age checks in Spirit Temple by splitting many of its regions into age-unknown and age-known variants. Also fixes the syntax error in Overworld.json and requires actually having the item when equip swapping.

I still want to fix potential keylocks from unrepeatable access to the hands, fix boss door access, and add more strats, particularly tricks in glitchless logic, strats using glitched damage, strats with dungeon shortcuts enabled, and supersliding to the Colossus hands from the arch. There are also parts where I don't know what the logic is expecting that I'd like to clarify or fix.

Hexocyte committed:
Correct glitch logic syntax error

Caused JSON parse error when generating on glitch logic.

Hexocyte committed:
Glitch logic: Require the item to equip swap it

Before it expected you to equip swap an item without having it.

Hexocyte commented:
Added strats
- https://youtu.be/1A7Axd8cbUs Superslide from Colossus HP to Silver Gauntlets
- https://youtu.be/ozXojJVvDCQ Explosiveless Mirror Shield with Hammer Recoil Hover Boost
- https://youtu.be/KPN87nva4Uk Colossus bean hammerboost
- https://youtu.be/9qPO9murMig Hoverboosts from Colossus hands to arch

Hexocyte committed:
Glitched: trickify entering Spirit via the hands

Hexocyte committed:
Glitched: add ground clips to Silver Gauntlets Hand

Hexocyte committed:
Glitched: adjust hovering in Colossus

Requires "All Uses Enabled" for hovering to the PoH and glitched damage
hovering from the arch to the hands without an item for contortion
hovering because of the length of these hovers.

I accounted for empty jumpslash instead of QPA for glitched damage,
which removes requirements for sticks and a cutscene item. Glitched
damage hovering now requires bombs. Glitched damage hovering to the
overhang requires both bombs and chus: bombs for hovering and a chu to
unlock the camera.

For brevity I removed hovers to Silver Gauntlets Hand that are redundant
with hovering to Mirror Shield Hand and then to Silvers.

Bombchu hovering without hoverboots is now expected when hoverboots
don't provide significant benefit.

I added a comment with how many explosives variations of hovering from
the arch to Mirror Shield Hand takes to document the exception to the
10 explosive limit.

Hexocyte committed:
Glitched: rename region to Spirit Temple Top

From Spirit Temple Beyond Final Locked Door, for brevity

Hexocyte committed:
Glitched: adjust hoverboots tech in Colossus

Added hoverboosts from the bean, crossing from Silver to Mirror, and to
the arch and the superslide from the arch.

Replaced mega with superslide/HESS for crossing between the hands.

Made changes in the temple to account for the possibility that the
hammerboosts add of reaching Mirror without explosives.

Hexocyte committed:
Glitched: trickify statue slide and statue climb

Add a new glitch toggle encompassing crossing from child side Spirit to
adult side via the statue and clipping into the head from adult side.

Hexocyte committed:
Glitched: add empty jumpslash & QPA to Spirit

Including shieldless Map Chest

Updated the name and description for glitch_qpa to include empty
jumpslash.

Hexocyte committed:
Glitched: fix some potential softlocks in Spirit

Fix some potential softlocks from unrepeatable access:
A. Access to Central Chamber with consumable bombchus or a losable
   bottle item or a losable shield.

But I didn't fix softlocks from:
B. Using hoverboots to climb the 1st silver block, hovering to the
   shortcut, and clipping past the silver blocks
C. Glitches that are never in logic
D. Hovering from OoB floor in the lobby to the shortcut
E. Unrepeatable access to the temple or Colossus in ER

A alone makes the logic more complicated. More would be too much or too
strict I think. Glitchless logic also has rare potential keylocks from
E, as stated in its Spirit Temple.json.

Hexocyte committed:
Glitched: redesign SpT key logic, harden age checks

Since opening any locked door can provide access to Central Chamber,
you only need 1 key to reach it with enough access and items.

There were many age checks in the logic that didn't ensure that the
proper age has access after the other age unlocks doors where the
proper age can't reach. For example if, with no optional glitches
enabled, there are 2 available keys, adult could unlock the adult-side
door to Central Chamber and another door past that to reach a hand, but
child would still be expected to go through the child side door to
boomerang the gold skulltula before the child-side knuckle.

There were also some uses of has_projectile(either) and
has_projectile(adult), which don't check for age, so for example the
player could be expected to use slingshot to hit the crystal switch in
Early Adult Spirit Temple when child has no access to the temple.

I hardened the age checks by splitting several regions into known- and
unknown-age variants. Location rules in ageless regions check that
either both ages have the necessary items or one age is at the strict
variant with the necessary items.

I added new helpers for the age-unknown logic. I also added the jumps
from the statue hands to the corner platforms (which are a trick in
glitchless logic) and fixed the Child Climb chests so that they would
always require projectiles.

Hexocyte committed:
Glitched: add more connections to adult Spirit

From before the locked doors to Adult Central Chamber Unknown, bypassing
the age-unknown logic in Central Chamber Unknown.

Hexocyte committed:
Glitched: reorder regions in Spirit

Put all the split regions together; make it closer to a normal
playthrough while keeping regions with open connections together.

Hexocyte committed:
Glitched: add more possibilities in Spirit

* Bomb pushes for silver block skip and crawlspace skip
* Hoverboots forward sidehop jumpslash to Compass Chest
* Megaflip to the central chamber GS
* Weirdshot to Child Spirit to skip the 2nd crawlspace in reverse
* Boulder room GS by clipping and jumping into the room behind it and
  getting the token with hookshot. Previously it expected hoverboots,
  and the hookshot requirement was missing.
* Boulder room GS with rang

Hexocyte committed:
Glitched: update logic for reaching Twinrova

Spirit Temple boss door and BK skip logic. The new logic considers
statue climb and head clip with no additional items and boss key skip by
Iron Boots megaflip or hoverboots.

Do not expect a damage ground clip BK skip when boss rooms are shuffled.
With damage multipliers it could leave insufficient health to defeat
Volvagia with a fairy, and there could be other issues if mixed pools
boss shuffle is added.

Also changed the exit for the elevator from the top to central chamber
to lead to adult side and removed skipping the elevator with hookshot.
It might've been expecting a weirdshot, but there's nothing in hookshot
range. The connection was redundant in the original logic but not
anymore with the hammerboosts in Colossus.

With these changes, the boss door access from the top is redundant, so
I removed the connection.

Added a description to the Spirit BK skip glitch toggle.

Hexocyte committed:
Glitched: restrict adult Spirit adult side

Reaching adult side from central chamber as adult now requires hookshot,
hoverboots, explosive clip for statue climb, or hovering. Needed because
of the shortcuts setting.

Hexocyte committed:
Glitched: add Colossus access w 4 keys & explosives

Add connection from Spirit Temple Central Chamber Adult Side Strict to
Desert Colossus and make the PoH reachable via it with a megaflip, a
hoverboost, or a hoverboots superslide. Also remove redundant
possibilities from the PoH access rule.

Hexocyte committed:
Glitched: revise Early Child Spirit

Split the part across the gaps into a new region. Rang was in logic as
an alternative to nuts for stunning the Armos. I removed that because
child has nuts available from the crate. Also, I think it was assumed
that adult would already have an item for crossing the gaps. But with
the TSC+crawlspace in logic, that's no longer the case; fixed. I added
restrictions for the nut crate as adult and duplicated it to Child
Climb with alternative restrictions so that it can be gotten by coming
from Central Chamber.

Add missing item and glitch toggle requirements to skip the crawlspaces.
I'm guessing the 2nd crawlspace with hoverboots is supposed to be EPG
and hookshot clip. I added the possibility of EPG there with explosives.
Also added EPG to the description of the actor glitch/EPG glitch toggle.

Hexocyte committed:
Glitched: add alternatives to damage in Spirit

After killing the child climb GS with a projectile, roll after jumping
to the token to avoid fall damage. Can also hover off of it or do an
iron boots dive (new glitch toggle).

For BK Chest, opening it with the flames up only loses a quarter heart
if you have roll invincibility for the first hit and put out the flames
by slashing. Added fairy or hookshot+nayrus to open the flaming chest,
slingshot to spawn the platform, and hovering to reach the switch
without the platform.

I considered removing the damage logic for BK Chest so that OHKO players
could be required to get a game over after getting the item, but I
figure that'd be a little tedious, so it's not worth doing.

Hexocyte commented:
* https://youtu.be/5wA-ZpG2pa8 "Spirit Temple Statue Climb Strat (Child Side)"
* https://youtu.be/yISWSywtdZQ "Spirit Temple Statue Slide as Adult"
* https://youtu.be/cGr5byNZSaI "Spirit Temple Explosiveless Statue Climb Setup (from Statue Arm)"
* https://youtu.be/vu8XiTp4Uuc "Setup for Spirit Temple Head Clip with Kokiri Boots"
* https://youtu.be/CuMTU-iYORE "Spirit Temple Statue Head Clip BK Skip Setup"
* https://youtu.be/hUC9xV_ep9I "Spirit Temple Bomb Push Silver Block Skip setup"
* https://youtu.be/WOJgQViboyY "spirit temple SoT Block Skull without hover boots"
* https://youtu.be/DRY3o1Gw7Y4 "Spirit temple compass with only hover boots"
* https://youtu.be/vQoV8g8YJSk "iron boots ledge dive, bomb OI dive"

Hexocyte committed:
Glitched: add 3-key access to Colossus hands

Also add some redundant 4-key access to make it easier to follow.

DuskTheUmbreon commented on data/Glitched World/Overworld.json:
```diff
@@ -609,17 +609,36 @@
         "time_passes": true,
-        "locations": {
+        "events": {
+            "Adult Spirit Hover Is Repeatable": "
```
```suggestion
            # Special logic is needed to make sure that it can be reached again after being accessed once.
            "Adult Spirit Hover Is Repeatable": "
```

DuskTheUmbreon replied:
Not happy with my wording on this. Might come back to it later.

Hexocyte replied:
I added "Used in Spirit logic, checks that adult can return to the hands after either age reaches them once from here"

Hexocyte committed:
Glitched: detrickify iron boots dive

Only used for 1 GS in Spirit with OHKO, and cutscene dives are generally
in base logic. When I added it, I thought it'd also be useful for LH GS
Lab Crate, but for that, you can also do the same ISG cutscene dive
that's in logic for Lab Dive, except with iron boots equipped.

Hexocyte committed:
Glitched: add more Colossus comments

One explaining the Adult Spirit Hover Is Repeatable event and another
explaining excluding glitchless Spirit hover strats

Hexocyte committed:
Glitched: require explosives for Spirit top elevator

DuskTheUmbreon commented:
Everything looks good. I'll make a few stylistic changes later probably, but this saved me a LOT of time and headache.
